app/fileUploadParsers/AnnualAllowanceParser.scala

```diff
@@ -17,26 +17,101 @@
 package fileUploadParsers
 
 import com.google.inject.Inject
+import config.FrontendAppConfig
 import forms.MemberDetailsFormProvider
-import models.UserAnswers
-import pages.chargeE.MemberDetailsPage
+import forms.chargeE.ChargeDetailsFormProvider
+import forms.mappings.Constraints
+import models.MemberDetails
+import models.chargeE.ChargeEDetails
+import pages.chargeE.{ChargeDetailsPage, MemberDetailsPage}
+import play.api.data.Form
+import play.api.data.validation.{Invalid, Valid}
+import play.api.i18n.Messages
+import play.api.libs.json.Json
 
-class AnnualAllowanceParser @Inject()(
-                                       memberDetailsFormProvider: MemberDetailsFormProvider
-                                     ) extends Parser {
+import java.time.LocalDate
 
+class AnnualAllowanceParser @Inject()(
+                                       memberDetailsFormProvider: MemberDetailsFormProvider,
+                                       chargeDetailsFormProvider: ChargeDetailsFormProvider,
+                                       config: FrontendAppConfig
+                                     ) extends Parser with Constraints {
+  //scalastyle:off magic.number
   override protected val totalFields: Int = 7
 
-  override protected def validateFields(ua: UserAnswers, index: Int, chargeFields: Array[String]): Either[ParserValidationErrors, UserAnswers] = {
-    val m = Map(
-      "firstName" -> firstNameField(chargeFields),
-      "lastName" -> lastNameField(chargeFields),
-      "nino" -> ninoField(chargeFields)
-    )
-    val form = memberDetailsFormProvider.apply()
-    form.bind(m).fold(
-      formWithErrors => Left(ParserValidationErrors(index, formWithErrors.errors.map(_.message))),
-      value => Right(ua.setOrException(MemberDetailsPage(index), value))
+  override protected def validHeader: String = config.validAnnualAllowanceHeader
+
+  private final object ChargeDetailsFieldNames {
+    val chargeAmount: String = "chargeAmount"
+    val dateNoticeReceivedDay: String = "dateNoticeReceived.day"
+    val dateNoticeReceivedMonth: String = "dateNoticeReceived.month"
+    val dateNoticeReceivedYear: String = "dateNoticeReceived.year"
+    val dateNoticeReceived: String = "dateNoticeReceived"
+    val isPaymentMandatory = "isPaymentMandatory"
+  }
+
+  private final object TaxYearErrorKeys {
+    val requiredKey = "annualAllowanceYear.fileUpload.error.required"
+    val invalidKey = "annualAllowanceYear.fileUpload.error.invalid"
+    val minKey = "annualAllowanceYear.fileUpload.error.past"
+    val maxKey = "annualAllowanceYear.fileUpload.error.future"
+  }
+
+  private def chargeDetailsValidation(startDate: LocalDate, index: Int, chargeFields: Array[String]): Either[Seq[ParserValidationError], ChargeEDetails] = {
+    val taxYearsErrors = validateTaxYear(startDate, index, chargeFields(3))
```

Comment:
would be better if you replace these number  ( 5 , 4 and 6 )  with constants.

Reply:
Done